### PR TITLE
Fix pathfinder bugs: returning nil frequently, broken A*, jump through solid nodes

### PR DESCRIPTION
**Pull request by Wuzzy2**
## Goal
Fix the biggest bugs of the pathfinder (`minetest.find_path`).

## Changes
* Fix failure if start or end pos are hovering a few nodes above air (#1701):
    * NOTE: I consider this to be an **important bugfix**. #1701 causes `minetest.find_path` to fail very frequently in common scenarios
    * If startpos is over air, pathfinder will first simulate a drop of `maxdrop` to surface. If surface was found, pos above that will become the start of the algorithm 
    * If destinationpos is over air, pathfinder will also simulate a drop, but with using `maxjump` as limit. If surface was found, pos above that will become the end of the algorithm
    * If startpos and/or destinationpos have been manipulated by the method above, the original startpos/destinationpos will be preserved in the end result, so there will be no ugly surprises on the Lua side
* Fix jumping going through solid matter
* Complete new implementation of `A*` / `A*_noprefetch`. It is now the *real* A* algorithm, not the fake one. The hilariously inefficient paths of Fake A* are gone
* Update documentation of `minetest.find_path`

## Notes
The so-called A* algorithm was completely bonkers (#9331) and also it wasn't even real A*. The offending function was `updateCostHeuristic`. This function is supposedly A* but it's actually not A* at all, it's a greedy recursive non-optimal algorithm, there's not even a concept of an open or closed list. If I understood the code correctly, neighbor nodes are usually just expanded immediately instead of putting them into a list. That's totally not A*. Often generating sometimes very long-winded paths instead of the optimal path. So yeah, completely bonkers.
Thus I decided to replace the function completely.

The new implementation of A* as well as `buildPath` is non-recursive (is there even a recursive A*?). The only function that is still recursive is `updateCostAll` (=Dijkstra).

Question: What is better?
Using the current heuristic function, that is, Manhattan distance on XZ plane only?
Or is it Manhattan distance on all XYZ coords?
In my tests, both function properly, but XYZ has tendency to go up and down on difficult terrain.
Currently, I did not touch the heuristic function.

## Resolved issues (hopefully)
* #1701
* #4198
* #9331 (@SmallJoker: this one is currently NOT on the “list of pathfinder issues” issue #8642)

## To do

It seems to work nicely so far, but needs more testing.

The code is not well-optimized and cleaned up yet, I appreciate any review.
I plan to switch the data structures of the A* search to boost performance, and maybe also seek out redundancies …
Currently, the way how `updateCostHeuristic` keeps track of open and closed list is kind of stupid by redundantly saving a boolean, but it works. I haven't figured out how to do it properly. `openList` is a set, but I think it should be a priority queue.

Known old bug (not caused by PR): Start/end position is one lower than it should be if Y < 0. Probably bad rounding error?

Please do NOT request any features, or this PR will blow up. Direct such comments to one of the pathfinder issues instead, or open a new issue. Concentrate on bugs in this PR, oddities, bad code, bad efficiency, etc.

After this PR, there's obviously still room for new fancy features. But before we even think about them, pathfinder must first become stable, that's very important.

## How to test

Download the `devtest` “game” and use the Pathfinder Tester (`testpathfinder:testpathfinder`) to play around with the pathfinder. Try everything to break it as hard as you can. ;)
Try to reproduce the listed issues. Test `A*` and `A*_noprefetch` extensively. Compare behaviour of previous version.

Try to see what happens if …
* goal or target or both are buried
* goal or target or both are above air (but within `maxdrop`/`maxjump` distance)
* you build a structure that provokes jumping, but something's in the way
* you are creative with your tests ;-)

Usage of Pathfinder Tester: Rightclick places target, punching searches path, sneak-punching changes. Note that jump height, drop height, search distance are hardcoded, so you need to get your hands dirty to change them.

### CODE REVIEW

**Comment by Wuzzy2** [comment]
Also: PLEASE ignore any coding style issues for now. I don't care (for now). I first want to make sure the code *works* and is *stable* and (more or less) *efficient*. When that is done and the PR is reasonably stable, you can attack me for my coding style all day long. :-)

**Comment by rubenwardy** [comment]
> Also: PLEASE ignore any coding style issues for now. I don't care (for now). I first want to make sure the code works and is stable and (more or less) efficient. 

This is how code review should be done :)

**Comment by Df458** [comment]
Re: heuristic, have you tried XYZ Manhattan but with a little bit of additional cost added to the Y distance? I can't guarantee that it'll work, but logically that should make the pathfinder prefer less-rough paths.

**Comment by Wuzzy2** [comment]
I have experimented with XYZ Manhattan a bit, but I still think XZ Manhattan is the most sensible for now.

Anyway, this is optional, this PR is a BUGFIX PR!

**Comment by rubenwardy** [comment]
This is the sort of code that *really* should be unit tested, and shouldn't be too hard to unit test

**Comment by Wuzzy2** [comment]
What exactly should I unit-test, tho?

**Comment by Wuzzy2** [comment]
Status report:

I have tried to improve performance by swiching from `std::set` to `str:priority_queue` for the open list, but I failed. What's worse, I have lost motivation. Note the priority queue is NOT required for this to work, it's a performance thing only. Getting pathfinder to work *at all* is more important IMO.

How to proceed?

**Comment by paramat** [comment]
Wuzzy2, do not worry about unit testing, we do not expect you to know how to do that, i would not remotely know how to either.
I think rubenwardy is just commenting on how it would be good to unit test this in future, it is not a request for this PR. Unit tests are obviously not required for this to be merged.

**Comment by rubenwardy** [comment]
>  I think rubenwardy is just commenting on how it would be good to unit test this in future, it is not a request for this PR. Unit tests are obviously not required for this to be merged.

This is correct, as it would be a non-trivial amount of time to set up the unit tests. If obtaining a world model to test using would be possible in a few hours, then I would request this as unit tests are vital in ensuring that a program is bug free

**Comment by rubenwardy** [comment]
This code is atrocious (not you, before this PR too) and I'm not surprised that it's so incredibly buggy

**Comment by Wuzzy2** [comment]
Is my PR rejected?

**Comment by rubenwardy** [comment]
No lol, it's just difficult to review

**Comment by Wuzzy2** [comment]
News!
I have successfully switched the open list data structure from `std::set` to `std::priority_queue`. This has significantly boosted performance (I checked).

I also did a bit of refactoring and added a lot of comments to explain the more difficult sections of the code. I hope it helps!

From my first post:

> Known old bug (not caused by PR): Start/end position is one lower than it should be if Y < 0. Probably bad rounding error?

A workaround is to call `minetest.find_path` only with rounded positions (`vector.round`). In any case, this is not an issue specific to `minetest.find_path`, this rounded position confusion occurs with many functions that expect an integer position. I consider this resolved now.

**Comment by Wuzzy2** [comment]
Can I increase my chances of getting this reviewed by fixing the coding style and restructuring the file?
Or should I wait with major code cleanups for after your review?

Note the price of a coding style change will be a very ugly diff.

Anyway, right now, I recommend to review the source file directly instead of the diff.

**Comment by sfan5** [comment]
I don't think there any benefit to fixing the code style in that file. Not now, at least.

File: src/pathfinder.cpp [hunk 1]
```diff
@@ -1006,194 +1119,166 @@ int Pathfinder::getXZManhattanDist(v3s16 pos)
...
-		DEBUG_OUT(LVL " Pathfinder: target found!" << std::endl);
-		return true;
-	}
-
-	bool retval = false;
+	// initial position
+	openList.push(source);
 
+	// the 4 cardinal directions
 	std::vector<v3s16> directions;
```
Comment (sfan5) [inline src/pathfinder.cpp]: this should be made `const static`

File: src/pathfinder.cpp [hunk 2]
```diff
@@ -1006,194 +1119,166 @@ int Pathfinder::getXZManhattanDist(v3s16 pos)
...
-	if (!g_pos.valid) {
-		ERROR_TARGET
-			<< LVL "Pathfinder: invalid next pos detected aborting" << std::endl;
-		return;
-	}
+	// The cost calculation should have set a source direction for all relevant nodes.
+	// To build the path, we go backwards from the destination until we reach the start.
+	unsigned int waypoints = 0;
+	while (true) {
+		waypoints++;
```
Comment (sfan5) [inline src/pathfinder.cpp]: `for(u32 waypoints = 1; waypoints++; ) {`

File: src/pathfinder.cpp [hunk 3]
```diff
@@ -1006,194 +1119,166 @@ int Pathfinder::getXZManhattanDist(v3s16 pos)
...
+			(down <= max_down)) {
+		testpos += v3s16(0, -1, 0);
+		down++;
+		node_at_pos = m_env->getMap().getNode(testpos);
+	}
+	//did we find surface?
+	if ((testpos.Y >= m_limits.MinEdge.Y) &&
+			(node_at_pos.param0 != CONTENT_IGNORE) &&
+			(ndef->get(node_at_pos).walkable)) {
+		if ((down - 1) <= max_down) {
```
Comment (sfan5) [inline src/pathfinder.cpp]: `down - 1` will wrap around when `down == 0`, most likely not intended

**Comment by Wuzzy2** [comment]
@sfan5: Done.

File: src/pathfinder.cpp [hunk 4]
```diff
@@ -929,19 +1042,21 @@ bool Pathfinder::updateAllCosts(v3s16 ipos,
...
+	};
 
-	directions.emplace_back(1,0, 0);
-	directions.emplace_back(-1,0, 0);
-	directions.emplace_back(0,0, 1);
-	directions.emplace_back(0,0,-1);
-
-	for (v3s16 &direction : directions) {
+	for (int d = 0; d <= 3; d++) {
+		v3s16 direction = directions[d];
```
Comment (sfan5) [inline src/pathfinder.cpp]: Pretty sure the `for (v3s16 &direction : directions)` works even with arrays

**Comment by sfan5** [comment]
I think it would make sense to add a `pathfinder_works_correctly` feature to builtin/game/features.lua so mods can detect that the pathfinder is usable again.

**Comment by Wuzzy2** [comment]
> Pretty sure the for (v3s16 &direction : directions) works even with arrays

No, it doesn't. But without `&`, it works (which I just did).

**Comment by Wuzzy2** [comment]
> (...) the pathfinder is usable again.

Given that the implementation of the A* algorithm was *completely* wrong, and given the nature of #1701 and #4198, I highly doubt the pathfinder was *ever* functional.

> I think it would make sense to add a pathfinder_works_correctly feature to builtin/game/features.lua

Is this your last feature request or will there be more?

**Comment by sfan5** [comment]
> Is this your last feature request or will there be more?

No this is all.

**Comment by Wuzzy2** [comment]
I'm not sure if I should really add something into `minetest.features`.

It feels kind of silly to basically use `minetest.features` to state that a given feature is *not* broken. If we start to do that for every feature that has been fixed, this table would explode.

To be clear: You were still able to call `minetest.find_path` without issue (=no crash), and it still found a path sometimes (but if often failed to find a path, when it should have found one, and returns nil), but it's rarely the optimal one. But that `minetest.find_path` returns `nil` if no path is found is an expected and documented behavior. What I mean with this: *This PR does not break API compability at all*!

I think it sets a really bad example if we start to teach modders to always check for `minetest.features` before calling `minetest.find_path`. This just blows up the code for no reason.

I'm waiting for the voices of other core devs.

**Comment by sfan5** [comment]
> You were still able to call minetest.find_path without issue

But if it returns nil you have no indication whether there is no possible path or the pathfinder is just still broken.

> What I mean with this: This PR does not break API compability at all!

(`minetest.features` is not a mechanism to ensure API compatibility)

> I think it sets a really bad example if we start to teach modders to always check for minetest.features before calling minetest.find_path

No, this isn't the idea. The idea is that a mod can use `find_path` instead of a possibly much more expensive Lua pathfinder, this is only possible if it can reliably detect if the pathfinder has been fixed yet.

**Comment by Wuzzy2** [comment]
After a quick chat, I have added now `minetest.features.pathfinder_works`.

File: src/pathfinder.cpp [hunk 5]
```diff
@@ -287,22 +278,32 @@ class Pathfinder {
...
+
+        /**
+         * go downwards from a position until some barrier
+         * is hit.
+         * @param pos position from which to go downwards
+         * @param max_down maximum distance to go downwards
+         * @return new position after movement; if too far down,
+                   pos is returned
+         */
+        v3s16         walkDownwards(v3s16 pos, unsigned int max_down);
```
Comment (SmallJoker) [inline src/pathfinder.cpp]: Please use tabs to indent, or tab + space for the comment above.

File: src/pathfinder.cpp [hunk 6]
```diff
@@ -378,6 +381,32 @@ class Pathfinder {
 #endif
 };
 
+/** Helper class for the open list priority queue in the A* pathfinder
+ *  to sort the pathfinder nodes by cost.
+ */
+class PathfinderCompareHeuristic
+{
+	private:
+		Pathfinder* myPathfinder;
```
Comment (SmallJoker) [inline src/pathfinder.cpp]: `Pathfinder *myPathfinder;`
Please correct all the placement of pointers and references in this class definition. Feel free to omit redundant brackets or one-liner statements.

File: src/pathfinder.cpp [hunk 7]
```diff
@@ -378,6 +381,32 @@ class Pathfinder {
...
+	public:
+		PathfinderCompareHeuristic(Pathfinder* pf) {
+			myPathfinder = pf;
+		}
+		bool operator() (v3s16 pos1, v3s16 pos2) {
```
Comment (SmallJoker) [inline src/pathfinder.cpp]: Please move the bracket to a new line.

File: src/pathfinder.cpp [hunk 8]
```diff
@@ -831,29 +924,49 @@ PathCost Pathfinder::calcCost(v3s16 pos, v3s16 dir)
...
+		MapNode node_target = m_env->getMap().getNode(targetpos);
+		MapNode node_jump = m_env->getMap().getNode(jumppos);
+		bool headbanger = false; // true if anything blocks jumppath
+
+		while ((node_target.param0 != CONTENT_IGNORE) &&
+				(ndef->get(node_target).walkable) &&
+				(targetpos.Y < m_limits.MaxEdge.Y)) {
+			//if the jump would hit any solid node, discard
+			if ((node_jump.param0 == CONTENT_IGNORE) ||
+				(ndef->get(node_jump).walkable)) {
```
Comment (SmallJoker) [inline src/pathfinder.cpp]: +1 indent for continued conditionals.

File: src/pathfinder.cpp [hunk 9]
```diff
@@ -1006,194 +1120,170 @@ int Pathfinder::getXZManhattanDist(v3s16 pos)
...
-			cost = calcCost(g_pos.pos, direction);
-			g_pos.setCost(direction, cost);
+/******************************************************************************/
+bool Pathfinder::updateCostHeuristic(v3s16 isource, v3s16 idestination)
+{
+	// A* search algorithm.
+
+	// The open list contains the pathfinder nodes that still need to be checked.
+	// The priority queue sorts the pathfinder nodes by estimated cost, with lowest cost on the top.
+	std::priority_queue<v3s16, std::vector<v3s16>, PathfinderCompareHeuristic> openList(PathfinderCompareHeuristic(this));
```
Comment (SmallJoker) [inline src/pathfinder.cpp]: Line and comment above are too long. please break where appropriate.

File: src/pathfinder.cpp [hunk 10]
```diff
@@ -1006,194 +1120,170 @@ int Pathfinder::getXZManhattanDist(v3s16 pos)
...
+			// update Y component of direction if neighbor requires jump or fall
+			v3s16 direction_3d = v3s16(direction_flat);
+			direction_3d.Y = cost.y_change;
+
+			// get position of true neighbor
+			v3s16 neighbor = current_pos + direction_3d;
+			v3s16 ineighbor = getIndexPos(neighbor);
+			PathGridnode& n_pos = getIndexElement(ineighbor);
+
+			if (cost.valid && (!n_pos.is_closed && !n_pos.is_open)) {
```
Comment (SmallJoker) [inline src/pathfinder.cpp]: Redundant parentheses, and misplaced reference above.

File: src/pathfinder.cpp (hunk 5 above)
Comment (sfan5) [reply src/pathfinder.cpp]: https://github.com/minetest/minetest/pull/9339#issuecomment-578248679

In fact I would suggest merging this without fixing the code style. The much-needed bugfixes are more important than the (still) wrong code style of the entire file.

**Comment by Wuzzy2** [comment]
@SmallJoker: Done.

**Pull request merged**